// casadistance/webpack.production.config.js
const { resolve } = require('path');
const ExtractTextPlugin = require('extract-text-webpack-plugin');
const base = require('./webpack.config');        // import the dev config

/* clone the base config so we can mutate it safely */
const config = { ...base };

/* 1. overwrite devtool / entry flags */
config.devtool = false;          // no source-maps in production

/* 2. replace style-loader → ExtractTextPlugin */
config.module.rules = config.module.rules.map(rule => {
  if (String(rule.test) === String(/\.scss$/)) {
    return {
      test: /\.scss$/,
      use: ExtractTextPlugin.extract({
        fallback: 'style-loader',
        use: ['css-loader', 'sass-loader'],
        publicPath: '../',
      }),
    };
  }
  return rule;
});

/* 3. add the plugin */
config.plugins = [
  ...config.plugins,
  new ExtractTextPlugin({ filename: 'style.css', allChunks: true }),
];

/* 4. update output path/name if you like */
config.output = {
  path: resolve(__dirname, 'dist'),
  filename: '[name]-bundle.js',
};

module.exports = config;
