// casadistance/webpack.production.config.js
const ManifestPlugin = require('webpack-manifest-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const CleanWebpackPlugin = require('clean-webpack-plugin')
const WebpackMd5Hash = require('webpack-md5-hash');
const CopyWebpackPlugin = require('copy-webpack-plugin');
const ExtractTextPlugin = require("extract-text-webpack-plugin");
const webpack = require('webpack');
const fs = require('fs');
const { resolve } = require('path');

/* babel */
const babelSettings = JSON.parse(fs.readFileSync(".babelrc"));

const config = {
	entry: {
		main: 
		[
			// 'babel-polyfill',
			'./src/scripts/main.js'
		],
	},


	output: {
		path: resolve(__dirname, 'dist'),
		filename: '[name]-bundle.js',
	},
	
    resolveLoader: {
		moduleExtensions: ["-loader"]
	},

	module: {
		noParse: /(mapbox-gl)\.js$/,
		rules: [
      	{
	        test: /\.scss$/,
	        exclude: /node_modules/,
	        use: ExtractTextPlugin.extract({
	          fallback: 'style-loader',
	          use: [
	            'css-loader',
	            {
	              loader: 'sass-loader',
	              query: {
	                sourceMap: false,
	                includePaths: [resolve(__dirname, "node_modules")],
	                data: null,
	              },
	            },
	          ],
	          publicPath: '../'
	        }),
	      },
		{
			test: /\.js$/,
			//exclude: /node_modules/,
			include: [
				resolve(__dirname, "src"),
			],
			use:[{
				loader: "babel-loader"
			}]
		},
		{
			test: /\.(jpe?g|png|giv|ttf|eot|svg|woff)(\?v=[0-9]\.[0-9]\.[0-9])?$/,
			exclude: /node_modules/,
			use: 'base64-inline-loader?name=[name].[ext]'
			/*use:[{
				loader: "file-loader",
				options: {
					emitFile: false
				}
			}]*/
		}],
	},

	plugins: [
		new webpack.DefinePlugin({
            'process.env': {
                'NODE_ENV': JSON.stringify(process.env.NODE_ENV)
            }
        }),
		new webpack.DefinePlugin({
			'process.env.BROWSER': JSON.stringify(true),
		}),
	    new WebpackMd5Hash(),
	 	new ManifestPlugin(),
	 	// new ExtractTextPlugin({ filename: './style.css', disable: false, allChunks: true }),
	 	new HtmlWebpackPlugin({
            template: './src/index.html', // input
            filename: 'index.html', // output (relative to output path)
        }),
        new CleanWebpackPlugin(['dist/*'], {verbose: true, watch: false}),
        new webpack.optimize.UglifyJsPlugin({
        	sourceMap: true,
            compress: {
				screw_ie8: true,
              	warnings: false,
  				comparisons: false,  // don't optimize comparisons
            }
        }),
    	new ExtractTextPlugin({ filename: './style.css', disable: false, allChunks: true })
	]

};

module.exports = config
